[DNM][WIP] client/leasemanager: create a package to build transaction…

ajwerner opened the pull request:
… scoped leases

This is a very WIP implementation of transaction scoped locks.
The interface is garbage and not very considered.

Nevertheless it seems to mostly work. The biggest problem I see so far is that
at high concurrency there are cases where a lock can lay down all of the
relevant intents and then still fail to commit. This happens because the TS
cache information is lost due to a lease transfer before the txn record is layed
down. Such a possibility is a big problem for the lease use cases which assume
that an intent will be valid until the transaction expires. To some extent this
problem makes sense given the transaction expiration isn't well defined before
the txn record has been written.

Release note: None

Commit:
[DNM][WIP] client/leasemanager: create a package to build transaction scoped leases

This is a very WIP implementation of transaction scoped locks.
The interface is garbage and not very considered.

Nevertheless it seems to mostly work. The biggest problem I see so far is that
at high concurrency there are cases where a lock can lay down all of the
relevant intents and then still fail to commit. This happens because the TS
cache information is lost due to a lease transfer before the txn record is layed
down. Such a possibility is a big problem for the lease use cases which assume
that an intent will be valid until the transaction expires. To some extent this
problem makes sense given the transaction expiration isn't well defined before
the txn record has been written.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45156)
<!-- Reviewable:end -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45156)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/kv/txn_coord_sender.go, line 882 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45156#-M0OxSfz4FNtz8X_jFPy:-M0OxSfz4FNtz8X_jFPz:b-v0so0a) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3490c087e87c1159b2f13916dcf10fb49910ea7/pkg/kv/txn_coord_sender.go#L882)):*
> ```Go
> 	defer tc.mu.Unlock()
> 	if tc.mu.txn.WriteTimestamp.Less(ts) {
> 		tc.mu.txn.WriteTimestamp = ts
> ```

does this ensure commit at >= `WriteTimestamp` even if the txn does no more writes?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45156)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/kv/txn_coord_sender.go, line 882 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45156#-M0OxSfz4FNtz8X_jFPy:-M0OxxzS503xyceVOiMg:b-szcyqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3490c087e87c1159b2f13916dcf10fb49910ea7/pkg/kv/txn_coord_sender.go#L882)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

does this ensure commit at >= `WriteTimestamp` even if the txn does no more writes?
</blockquote></details>

Yes. The timestamp for the `EndTxnRequest` comes from the `WriteTimestamp`


<!-- Sent from Reviewable.io -->

sumeerbhola committed:
[DNM][WIP] client/leasemanager: additions for leases/advisory locks

- Txn.ForceHeartbeat() writes a transaction record.
- Txn.ExpiryTimestamp() returns the current expiry timestamp.
- TODOs for plumbing a special Intent that updates the ts cache when removed due to abort.

Release note: None

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45156)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/kv/txn_interceptor_heartbeater.go, line 238 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45156#-M0QU7D749xj02wUeYRf:-M0QU7D90n5CBpUOKNbQ:b-wn4acw) ([raw file](https://github.com/cockroachdb/cockroach/blob/c263b8579e6534be5d235180d5d47246b7bab955/pkg/kv/txn_interceptor_heartbeater.go#L238)):*
> ```Go
> 
> func (h *txnHeartbeater) forceHeartbeat() error {
> 	h.forceHeartbeatCh <- struct{}{}
> ```

The synchronization here isn't clear. It seems like you can't call this more than once. Make that clear from the API?

I sort of also wish this call and its parents took a context and were cancelable. 

I think we can come up with fancier dances with channels to make this reentrant and deal with context cancelation. 


<!-- Sent from Reviewable.io -->

sumeerbhola committed:
[DNM][WIP] client/leasemanager: additions for leases/advisory locks

More plumbing for special intent that updates ts cache.
Remaining work is to somehow update the ts cache in cmd_resolve_intent.go
and cmd_resolve_intent_range.go,

Release note: None

sumeerbhola committed:
[DNM][WIP] client/leasemanager: additions for leases/advisory locks

Plumbing to update the ts cache when removing a "leasing intent".

Release note: None

sumeerbhola committed:
[DNM][WIP] client/leasemanager: additions for leases/advisory locks

Added proof sketch for lease/lock implementation.
Added description of alternative that does not have contention
between concurrent acquisitions of shared lease.

Release note: None
